stan/math/opencl/kernels/gp_exp_quad_cov.hpp

```diff
@@ -0,0 +1,101 @@
+#ifndef STAN_MATH_OPENCL_KERNELS_GP_EXP_QUAD_COV_HPP
+#define STAN_MATH_OPENCL_KERNELS_GP_EXP_QUAD_COV_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/kernel_cl.hpp>
+
+namespace stan {
+namespace math {
+namespace opencl_kernels {
+// \cond
+static const char* gp_exp_quad_cov_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * GPU part of calculation of squared exponential kernel.
+     *
+     * @param[in] x input vector or matrix
+     * @param[out] res squared distances between elements of x
+     * @param sigma_sq squared standard deviation
+     * @param neg_half_inv_l_sq -0.5 / square(length scale)
+     * @param size number of elements in x
+     * @param element_size the number of doubles that make one element of x
+     */
+    __kernel void gp_exp_quad_cov(__global double* x, __global double* res,
+                                  double sigma_sq, double neg_half_inv_l_sq,
+                                  int size, int element_size) {
+      const int i = get_global_id(0);
+      const int j = get_global_id(1);
+      if (i < size && j < (size - 1) && i > j) {
+        double sum = 0;
+        for (int k = 0; k < element_size; k++) {
+          double d = x[i * element_size + k] - x[j * element_size + k];
```

Comment:
double is getting created and destroyed every time here. I think it's fine to pull it out of the loop and set it to zero?

Though I would bet 20 bucks the compiler sees this and does something smart anyway

Reply:
I could, but I think the code is clearer the way it is. Creation and destruction of a double is no-op anyway.